language-server/docs/ExtensionnServices.md

```diff
@@ -0,0 +1,160 @@
+
+<a name="ComponentsOfAService"></a>
+## Components of a Service
+In order to register a new service there are three must have components. 
+1. Service Implementation
+2. Server Capability Registration
+3. Client Capability Registration
+
+<a name="ServiceImplementation"></a>
+### Service Implementation
+The service implementation defines the service namespace and the service operation. The service is dynamically picked and registered at the server startup. The services will be registered with Java SPI implementation. In order to register a new service, following approach should be considered and adding the following annotation is mandatory.
+```Java
+@JavaSPIService("org.ballerinalang.langserver.commons.service.spi.ExtendedLanguageServerService")
+```
+
+Eg: 
+```Java
+@JavaSPIService("org.ballerinalang.langserver.commons.service.spi.ExtendedLanguageServerService")
+@JsonSegment("exampleService")
+public class CustomService implements ExtendedLanguageServerService {
+    @JsonRequest
+    public CompletableFuture<CustomResponse> syntaxApiCalls(CustomRequestParams request) {
+        // This is a custom request implementation denoted by the @JsonRequest annotation
+    }
+
+    @JsonNotification
+    public void syntaxApiCalls(CustomNotificationParams request) {
+        // This is a custom notification implementation denoted by the @JsonNotification annotation
+    }
+}
+```
+
+As per the above snippet we can register the service and get it working without any other configuration associated with the service. 
+The Ballerina Lannguage Server exposes context information for the use of the extension APIs to be used. Depending on the requirement. the extension developer can use different context information as follows,
+
+#### Default/ without overriding `init` method
+This will ensure the default behaviour without exposing context information for the extension
+
+#### Initialization option-1
+```Java
+default void init(LanguageServer langServer,
+                    WorkspaceManager workspaceManager) {
+    // access and store the parameters
+}
+```
+Exposes the language server instance and the default workspace manager.
+
+#### Initialization option-2
+```Java
+default void init(LanguageServer langServer,
+                    WorkspaceManager workspaceManager,
+                    LanguageServerContext serverContext) {
+    // access and store the parameters
+}
+```
+Additionally exposes the `LanguageServerContext`
+
+#### Initialization option-3
+```Java
+default void init(LanguageServer langServer,
+                      WorkspaceManagerProxy workspaceManagerProxy,
+                      LanguageServerContext serverContext) {
+    // access and store the parameters
+}
+```
+Instead of the `WorkspaceManager` this exposes the `WorkspaceManagerProxy`. The `WorkspaceManagerProxy` will expose both the default `WorkspaceManager` and the `expr` file scheme based workapace manager.
```

Comment:
```suggestion
Instead of the `WorkspaceManager`, this exposes the `WorkspaceManagerProxy`. The `WorkspaceManagerProxy` will expose both the default `WorkspaceManager` and the `expr` file scheme based workapace manager.
```